packages/toc/src/toc_tree.tsx

```diff
@@ -65,10 +78,17 @@ class TOCTree extends React.Component<IProperties, IState> {
     // Map the heading objects onto a list of JSX elements...
     let i = 0;
     let list: JSX.Element[] = this.props.toc.map(el => {
+      const contextMenu = new Menu({ commands: this.props.commands });
```

Comment:
It might be better to use the existing `app.contextMenu` and `app.commands` instead of creating new `ContextMenu` and `CommandRegistry` objects

Reply:
Let me know if you need any help adding those or there was a reason you couldn't use those! You'll want to pass them to the `TableOfContents` object in the `activate` function in `packages/toc-extension/src/index.ts`.

Reply:
Hi @marthacryan, thanks so much for looking at the code! 
I actually couldn't find a way to update a command in the command registry. That's why I am creating a new `commands` when toc updates. Is there a way to solve this? (and sorry I missed this thread)

Reply:
Are you *updating* a command or *adding* a new command? You probably shouldn't override any commands, it's better to just create a new command. But if you're using a new command, it should work the same as how you're adding the command to the registry here. 

One thought that occurs to me is that maybe your code is creating this command registry and adding the command multiple times. If so, it should probably be moved so that it's only being added as a command once (usually commands are added in the `activate` function of an extension). If it needs to be specific to the ToC instance you can use the `args` to help with that. 

Let me know if that makes sense / you need any clarifications!

Reply:
@marthacryan Thanks for following up.
Things in `args` need to be serialized, that's why I struggled using a singleton commands (app.commands) and putting the toc instance in `args` in `contextMenu.addItem`. But, I found a workaround this. I made the list of headings a property of the TOC object, so now using app.commands is possible. 

However, for using app.contextMenu, I run into a problem. `app.contextMenu` is a Lumino ContextMenu and its `.open` method takes a `MouseEvent`. and I have not found a way to cast the `React.MouseEvent` to `MouseEvent`. I would say that a blocker at this point. 

I have also tried creating a singleton Lumino`Menu` for Toc, but again ran into some limitation. Because, Lumino `Menu` does not have a updateItem method, so when the every time a `position` of a heading updates in the Toc Tree, you would need to neither 1) loop through all the items in the Menu, find and delete the "run cell(s)" item or 2) nuke the whole `Menu`. IMHO neither is more elegant than creating a new `Menu` each time.

Reply:
Glad the command part is working! Nice work!

I just noticed that you're triggering the context menu opening using the `open` command - in other parts of JL usually the context menu is set up using a selector (see https://github.com/jupyterlab/jupyterlab/blob/a8dff39a1a2be850a80861e05145e763079b8794/packages/filebrowser-extension/src/index.ts#L380 for example) which triggers a command - so you'd just need to add a special class to the `li` instead of adding `onContextMenu`, and then you'd move that context menu `addItem` call to the same place as where you're adding the command. The issue there is that you'd need to pass the arg `position` and as far as I know, you wouldn't be able to pass any args to a context menu option, so it'd have to rely on the state of the ToC to find the "active" entry in the ToC. Currently, right-clicking on a ToC entry doesn't set it to the "active" entry, but that should be pretty straightforward to change.

Reply:
@marthacryan Thank you for the clear step-by-step:) let me try this, will let you know if it works